### ISSUE TITLE
Installation Fails for bash >= 4.4

### ISSUE THREAD

**Issue by Zalgo2462** [activecm/rita#188]
RITA relies on sourcing the bash rc file in order to grab details regarding previous installations (go details, rita's install location, etc).

In order to get around interactive checks in various bashrc files, the installer makes use of the "set -i" command [HERE](https://github.com/activecm/rita/blob/02b69e10c80518e7c9086fe75bd45e5f904d093e/install.sh#L31). However, the developers of bash have removed this option in version 4.4 [HERE](https://github.com/activecm/rita/blob/02b69e10c80518e7c9086fe75bd45e5f904d093e/install.sh#L3).

**Comment by Zalgo2462** [comment]
Here's a patch. Didn't PR due to large number of installer bugs/ oddities.

```diff --git a/install.sh b/install.sh
index bb18229..176673f 100755
--- a/install.sh
+++ b/install.sh
@@ -19,27 +19,11 @@ set -o errtrace
 set -o pipefail
 
 # Fix $HOME for users under standard sudo
+# Particulary, Ubuntu doesn't seem to need this
 if [ ! -z ${SUDO_USER+x} ]; then
 	HOME="$( getent passwd $SUDO_USER | cut -d: -f6 )"
 fi
 
-# Make sure to source the latest .bashrc
-# Hack the PS1 variable to get around ubuntu .bashrc
-_OLD_PS1=$PS1
-PS1=" "
-# Hack the interactive flag to get around other .bashrc's
-set -i
-# Make sure weirdness doesn't happen with autocomplete/ etc
-set -o posix
-
-source $HOME/.bashrc
-
-# Clean up our hacks
-set +o posix
-set +i
-PS1=$_OLD_PS1
-unset _OLD_PS1
-
 __help() {
 	__title
 
@@ -54,6 +38,12 @@ Options:
 HEREDOC
 }
 
+__source_rita_vars() {
+	if [ -f $_RITA_VARS ]; then
+		source $_RITA_VARS
+	fi
+}
+
 __explain() {
 	cat <<HEREDOC
 This script will:
@@ -117,9 +107,11 @@ __uninstall() {
 	fi
 	printf "\t[!] Removing $_CONFIG_PATH \n"
 	rm -rf "$_CONFIG_PATH"
+	printf "\t[!] Removing entries from $HOME/.bashrc \n"
+	sed -i "s|source $_RITA_VARS||g" $HOME/.bashrc
 }
 
-__setPkgMgr() {
+__set_pkg_mgr() {
 	# _PKG_MGR = 1: APT: Ubuntu 14.04, 16.04 and Security Onion (Debian)
 	# _PKG_MGR = 2: YUM: CentOS (Old RHEL Derivatives)
 	# _PKG_MGR = 3: Unsupported
@@ -226,7 +218,7 @@ __install_go() {
 		fi
 		printf "\t[+] Adding Go to the PATH...\n"
 		export PATH="$PATH:/usr/local/go/bin"
-		echo 'export PATH=$PATH:/usr/local/go/bin' >> $HOME/.bashrc
+		echo 'export PATH=$PATH:/usr/local/go/bin' >> $_RITA_VARS
 	else
 		printf "\t[+] Go is installed...\n"
 	fi
@@ -235,8 +227,8 @@ __install_go() {
 	if [ -z ${GOPATH+x} ]; then
 		( # Set up the GOPATH
 			mkdir -p $_INSTALL_PREFIX/{src,pkg,bin}
-			#echo "export GOPATH=$_INSTALL_PREFIX" >> $HOME/.bashrc
-			echo "export PATH=\$PATH:$_INSTALL_PREFIX/bin" >> $HOME/.bashrc
+			#echo "export GOPATH=$_INSTALL_PREFIX" >> $_RITA_VARS
+			echo "export PATH=\$PATH:$_INSTALL_PREFIX/bin" >> $_RITA_VARS
 		) & __load "\t[+] Configuring Go dev environment"
 		export GOPATH=$_INSTALL_PREFIX
 		export PATH=$PATH:$_INSTALL_PREFIX/bin
@@ -263,8 +255,8 @@ __install_bro() {
 
 	if [ ! $(command -v bro) ];	then
 		printf "\t[+] Adding Bro to the PATH...\n"
-		echo 'export PATH=$PATH:/opt/bro/bin' >> $HOME/.bashrc
-		PATH=$PATH:/opt/bro/bin
+		echo 'export PATH=$PATH:/opt/bro/bin' >> $_RITA_VARS
+		export PATH=$PATH:/opt/bro/bin
 	fi
 	chmod 2755 /opt/bro/logs
 }
@@ -289,7 +281,7 @@ __install() {
 
 	# Check if RITA is already installed, if so ask if this is a re-install
 	if [ ! -z $(command -v rita) ] \
-		|| [ -d /opt/rita ] \
+		|| [ -d $_INSTALL_PREFIX ] \
 		|| [ -d $_CONFIG_PATH ]
 	then
 		printf "[+] RITA is already installed.\n"
@@ -307,7 +299,7 @@ __install() {
 	__explain
 
 	# Figure out which package manager to use
-	__setPkgMgr
+	__set_pkg_mgr
 
 	# Update package sources
 	__freshen_packages
@@ -319,6 +311,10 @@ __install() {
 	# Determine the OS, needs lsb-release
 	__setOS
 
+	# Create the installation directories
+	mkdir -p $_INSTALL_PREFIX
+	mkdir -p $_CONFIG_PATH
+
 	if [[ "${_INSTALL_BRO}" = "true" ]]
 	then
 		__install_bro
@@ -348,7 +344,6 @@ __install() {
 
 
 	( # Install the base configuration files
-		mkdir -p $_CONFIG_PATH
 		cd $GOPATH/src/github.com/ocmdev/rita
 		cp ./LICENSE $_CONFIG_PATH/LICENSE
 		cp ./etc/rita.yaml $_CONFIG_PATH/config.yaml
@@ -360,6 +355,9 @@ __install() {
 		chmod 666 $_CONFIG_PATH/safebrowsing
 	) & __load "[+] Installing config files to $_CONFIG_PATH"
 
+	# Load the $_RITA_VARS in the bashrc
+	echo "source $_RITA_VARS" >> $HOME/.bashrc
+
 	echo -e "
 In order to finish the installation, reload your bash config
 with 'source ~/.bashrc'. Make sure to configure Bro and run
@@ -374,12 +372,12 @@ to stop MongoDB, run 'sudo service mongod stop'."
 }
 
 # start point for installer
-__entry() {	
+__entry() {
 	_INSTALL_BRO=true
 	_INSTALL_MONGO=true
 	_INSTALL_PREFIX=/opt/rita
 	_CONFIG_PATH=/etc/rita
-	_INSTALL_RITA=true	
+	_INSTALL_RITA=true
 	_UNINSTALL_RITA=false
 
 	# Parse through command args
@@ -396,10 +394,10 @@ __entry() {
 				_INSTALL_BRO=false
 				_INSTALL_MONGO=false
 				;;
-			--disable-bro) 
+			--disable-bro)
 				_INSTALL_BRO=false
 				;;
-			--disable-mongo) 
+			--disable-mongo)
 				_INSTALL_MONGO=false
 				;;
 			--prefix)
@@ -412,6 +410,10 @@ __entry() {
 	  shift
 	done
 
+	# Load rita installation details if they exist
+	_RITA_VARS="$_CONFIG_PATH/rita_vars.sh"
+	__source_rita_vars
+
 	# Check to see if the user has permission to install RITA
 	if __checkPermissions
 	then
@@ -419,7 +421,7 @@ __entry() {
 		then
 			__uninstall
 			exit 0
-		fi		
+		fi
 		if [[ "${_INSTALL_RITA}" = "true" ]]
 		then
 			__install
@@ -428,8 +430,6 @@ __entry() {
 	else
 		printf "You do NOT have permission install RITA\n\n"
 	fi
-
-
 }
 
 __entry "${@:-}"
```

**Comment by Zalgo2462** [comment]
This patch creates issues of its own. 

# Installer Wrongly Adds Go to the PATH
Sudo strips all environment variables including the PATH. When a user invokes the install script with sudo, /usr/local/go/bin/go may be in their PATH. However, the script cannot see this as sudo has replaced the PATH. This causes the script to add go to the bashrc even though it may already be there.

https://github.com/activecm/rita/blob/02b69e10c80518e7c9086fe75bd45e5f904d093e/install.sh#L219

The best way to handle this may be to prompt the user if go isn't in the usual location asking whether they would like to install go or provide a different install location. 

Realistically, go *should* be installed in its usual location. If go is installed in a different location, GOROOT must also be set. 

# Installer Wrongly Adds Bro to the PATH
Same as above.

# Installer Wrongly Creates a new GOPATH
The GOPATH is also stripped under sudo.

**Comment by Zalgo2462** [comment]
The solution may be to self sudo the script.

Test if SUDO_USER is set and _SELF_SUDO not set: Warn and Exit.

If EUID != 0 (ie not root): run this script with `sudo PATH=$PATH GOPATH=$GOPATH _SELF_SUDO=true "$0" "$@"`

**Comment by Zalgo2462** [comment]
Patch with the above solution:

```diff --git a/install.sh b/install.sh
index bb18229..6d195de 100755
--- a/install.sh
+++ b/install.sh
@@ -3,6 +3,30 @@
 # RITA is brought to you by Offensive CounterMeasures.
 # offensivecountermeasures.com
 
+# Self sudo gadget to preserve needed environment variables
+# Prevent user running sudo themselves
+if [[ ! -z ${SUDO_USER+x} && -z ${_SELF_SUDO+x} ]]; then
+	printf "Please run the RITA installer without sudo.\n"
+	exit
+fi
+# If not root, run as sudo
+if [[ $EUID -ne 0 ]]; then
+	printf "The installer will now attempt to elevate priveledges with sudo...\n"
+	sudo -v
+	if [[ $? -ne 0 ]]; then
+		printf "Unable to elevate privledges with sudo.\n"
+		exit 1
+	fi
+	exec sudo PATH=$PATH GOPATH=$GOPATH _SELF_SUDO="TRUE" "$0" "$@"
+fi
+
+# Fix $HOME for users under sudo
+# Particulary, Ubuntu doesn't seem to need this
+if [ ! -z ${SUDO_USER+x} ]; then
+	HOME="$( getent passwd $SUDO_USER | cut -d: -f6 )"
+fi
+
+# Constants
 _NAME=$(basename "${0}")
 _FAILED="\e[91mFAILED\e[0m"
 _SUCCESS="\e[92mSUCCESS\e[0m"
@@ -18,28 +42,7 @@ set -o errexit
 set -o errtrace
 set -o pipefail
 
-# Fix $HOME for users under standard sudo
-if [ ! -z ${SUDO_USER+x} ]; then
-	HOME="$( getent passwd $SUDO_USER | cut -d: -f6 )"
-fi
-
-# Make sure to source the latest .bashrc
-# Hack the PS1 variable to get around ubuntu .bashrc
-_OLD_PS1=$PS1
-PS1=" "
-# Hack the interactive flag to get around other .bashrc's
-set -i
-# Make sure weirdness doesn't happen with autocomplete/ etc
-set -o posix
-
-source $HOME/.bashrc
-
-# Clean up our hacks
-set +o posix
-set +i
-PS1=$_OLD_PS1
-unset _OLD_PS1
-
+# Helper functions
 __help() {
 	__title
 
@@ -54,6 +57,12 @@ Options:
 HEREDOC
 }
 
+__source_rita_vars() {
+	if [ -f $_RITA_VARS ]; then
+		source $_RITA_VARS
+	fi
+}
+
 __explain() {
 	cat <<HEREDOC
 This script will:
@@ -117,9 +126,11 @@ __uninstall() {
 	fi
 	printf "\t[!] Removing $_CONFIG_PATH \n"
 	rm -rf "$_CONFIG_PATH"
+	printf "\t[!] Removing entries from $HOME/.bashrc \n"
+	sed -i "s|source $_RITA_VARS||g" $HOME/.bashrc
 }
 
-__setPkgMgr() {
+__set_pkg_mgr() {
 	# _PKG_MGR = 1: APT: Ubuntu 14.04, 16.04 and Security Onion (Debian)
 	# _PKG_MGR = 2: YUM: CentOS (Old RHEL Derivatives)
 	# _PKG_MGR = 3: Unsupported
@@ -226,17 +237,17 @@ __install_go() {
 		fi
 		printf "\t[+] Adding Go to the PATH...\n"
 		export PATH="$PATH:/usr/local/go/bin"
-		echo 'export PATH=$PATH:/usr/local/go/bin' >> $HOME/.bashrc
+		echo 'export PATH=$PATH:/usr/local/go/bin' >> $_RITA_VARS
 	else
 		printf "\t[+] Go is installed...\n"
 	fi
 
-	# Check if the GOPATH isn't set
-	if [ -z ${GOPATH+x} ]; then
+	# Check if the GOPATH isn't set or empty via self-sudo
+	if [ -z $GOPATH ]; then
 		( # Set up the GOPATH
 			mkdir -p $_INSTALL_PREFIX/{src,pkg,bin}
-			#echo "export GOPATH=$_INSTALL_PREFIX" >> $HOME/.bashrc
-			echo "export PATH=\$PATH:$_INSTALL_PREFIX/bin" >> $HOME/.bashrc
+			#echo "export GOPATH=$_INSTALL_PREFIX" >> $_RITA_VARS
+			echo "export PATH=\$PATH:$_INSTALL_PREFIX/bin" >> $_RITA_VARS
 		) & __load "\t[+] Configuring Go dev environment"
 		export GOPATH=$_INSTALL_PREFIX
 		export PATH=$PATH:$_INSTALL_PREFIX/bin
@@ -263,8 +274,8 @@ __install_bro() {
 
 	if [ ! $(command -v bro) ];	then
 		printf "\t[+] Adding Bro to the PATH...\n"
-		echo 'export PATH=$PATH:/opt/bro/bin' >> $HOME/.bashrc
-		PATH=$PATH:/opt/bro/bin
+		echo 'export PATH=$PATH:/opt/bro/bin' >> $_RITA_VARS
+		export PATH=$PATH:/opt/bro/bin
 	fi
 	chmod 2755 /opt/bro/logs
 }
@@ -289,7 +300,7 @@ __install() {
 
 	# Check if RITA is already installed, if so ask if this is a re-install
 	if [ ! -z $(command -v rita) ] \
-		|| [ -d /opt/rita ] \
+		|| [ -d $_INSTALL_PREFIX ] \
 		|| [ -d $_CONFIG_PATH ]
 	then
 		printf "[+] RITA is already installed.\n"
@@ -307,7 +318,7 @@ __install() {
 	__explain
 
 	# Figure out which package manager to use
-	__setPkgMgr
+	__set_pkg_mgr
 
 	# Update package sources
 	__freshen_packages
@@ -319,6 +330,10 @@ __install() {
 	# Determine the OS, needs lsb-release
 	__setOS
 
+	# Create the installation directories
+	mkdir -p $_INSTALL_PREFIX
+	mkdir -p $_CONFIG_PATH
+
 	if [[ "${_INSTALL_BRO}" = "true" ]]
 	then
 		__install_bro
@@ -348,7 +363,6 @@ __install() {
 
 
 	( # Install the base configuration files
-		mkdir -p $_CONFIG_PATH
 		cd $GOPATH/src/github.com/ocmdev/rita
 		cp ./LICENSE $_CONFIG_PATH/LICENSE
 		cp ./etc/rita.yaml $_CONFIG_PATH/config.yaml
@@ -360,6 +374,9 @@ __install() {
 		chmod 666 $_CONFIG_PATH/safebrowsing
 	) & __load "[+] Installing config files to $_CONFIG_PATH"
 
+	# Load the $_RITA_VARS in the bashrc
+	echo "source $_RITA_VARS" >> $HOME/.bashrc
+
 	echo -e "
 In order to finish the installation, reload your bash config
 with 'source ~/.bashrc'. Make sure to configure Bro and run
@@ -374,12 +391,12 @@ to stop MongoDB, run 'sudo service mongod stop'."
 }
 
 # start point for installer
-__entry() {	
+__entry() {
 	_INSTALL_BRO=true
 	_INSTALL_MONGO=true
 	_INSTALL_PREFIX=/opt/rita
 	_CONFIG_PATH=/etc/rita
-	_INSTALL_RITA=true	
+	_INSTALL_RITA=true
 	_UNINSTALL_RITA=false
 
 	# Parse through command args
@@ -396,10 +413,10 @@ __entry() {
 				_INSTALL_BRO=false
 				_INSTALL_MONGO=false
 				;;
-			--disable-bro) 
+			--disable-bro)
 				_INSTALL_BRO=false
 				;;
-			--disable-mongo) 
+			--disable-mongo)
 				_INSTALL_MONGO=false
 				;;
 			--prefix)
@@ -412,24 +429,20 @@ __entry() {
 	  shift
 	done
 
-	# Check to see if the user has permission to install RITA
-	if __checkPermissions
+	# Load rita installation details if they exist
+	_RITA_VARS="$_CONFIG_PATH/rita_vars.sh"
+	__source_rita_vars
+
+	if [[ "${_UNINSTALL_RITA}" = "true" ]]
 	then
-		if [[ "${_UNINSTALL_RITA}" = "true" ]]
-		then
-			__uninstall
-			exit 0
-		fi		
-		if [[ "${_INSTALL_RITA}" = "true" ]]
-		then
-			__install
-			exit 0
-		fi
-	else
-		printf "You do NOT have permission install RITA\n\n"
+		__uninstall
+		exit 0
+	fi
+	if [[ "${_INSTALL_RITA}" = "true" ]]
+	then
+		__install
+		exit 0
 	fi
-
-
 }
 
 __entry "${@:-}"

```

**Comment by Zalgo2462** [comment]
Alternative solutions: build RITA in docker; only sudo for whats needed (with the correct environment passing) and preventing users from running the script with sudo in the first place.

**Comment by ethack** [comment]
Fixed with #197
